File: Two Pointers/Valid-Palindrome-II.js
/**
 * @param {string} s
 * @return {boolean}
 */
const validPalindrome = (s) => {
  if (s.length === 1) {
    return true;
  }

  let leftPointer = 0;
  let rightPointer = s.length - 1;

  while (leftPointer < rightPointer) {
    const isSame = s[leftPointer] === s[rightPointer];

    if (!isSame) {
      return (
        isPalindrome(s.slice(leftPointer + 1, rightPointer + 1)) ||
        isPalindrome(s.slice(leftPointer, rightPointer))
      );
    }

    leftPointer++;
    rightPointer--;
  }

  return true;
};

const isPalindrome = (s) => {
  if (s.length === 1) {
    return true;
  }

  let leftPointer = 0;
  let rightPointer = s.length - 1;

  while (leftPointer < rightPointer) {
    const isSame = s[leftPointer] === s[rightPointer];

    if (!isSame) {
      return false;
    }

    leftPointer++;
    rightPointer--;
  }

  return true;
};

File: Go/Valid-Palindrome-II.go
package main


func validPalindrome(s string) bool {
	if len(s) == 1 {
		return true

	left, right := 0, len(s)-1

	for left < right {
		if s[left] != s[right] {
			return isPalindrome(s[left+1:right+1]) || isPalindrome(s[left:right])
		}

		left++
		right--
	}
	return true
}

func isPalindrome(s string) bool {
	if len(s) == 1 {
		return true
	}

	left, right := 0, len(s)-1

	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}

	return true
}